### PR TITLE
feat: det deploy local generates a password for you [DET-10197]

### PR DESCRIPTION
**Pull request by jesse-amano-hpe**
## Ticket
[DET-10197]

## Description

BREAKING CHANGE: if an initial user password is not specified, the `det deploy local` command generates one automatically; if that random password is recognized as the correct one for `determined`, it is printed so users can log in with it and change their password. Otherwise, it is assumed that this is a relaunch of a cluster with an existing database.

The proposal for this change was first described here: https://hpe-aiatscale.slack.com/archives/CLCE8D998/p1716571996243779?thread_ts=1716404697.160779&cid=CLCE8D998

### Additional notes
`det deploy aws` and `det deploy gcp` can rely on Terraform state to determine whether a database (or database image) already contains passwords, and prompt interactively if not, and `helm` can run some simple checks during templating. However, `det deploy local` sets up a database in a docker container based on a volume that may or may not be empty, and it's hard to know migration states etc. until master is healthy. Rather than rely on `master` to decide in this case (which has caused problems in the past), this solution guarantees there will always be a password set on cluster creation, with the only major downside being that the admin user is still responsible for changing their password if the deployment logs are somehow exposed long-term.

## Test Plan
I've manually tested and added integration tests, but it is probably worth triple-checking that the following is still true on next release:
- `det deploy local master-up --delete-db` prompts for the user to enter an initial password, displays requirements if the first attempt is blank or weak, and falls back to displaying a generated password if the user can't set this correctly.
- `det deploy local master-up` (with an existing database volume) allows login with whatever passwords were set before and does not print the generated password message
- `det deploy local master-up --delete-db --initial-user-password=$PASSWORD` sets up the provided passwords and does not print the generated password message
- `time det deploy local master-up --det-version='203370a6cdb048aaa579c7406f584bb428e76df4' --master-port=8080 --delete-db --initial-user-password '' &` (runs detached in background) skips prompting for a password due to the detached TTY, and displays the generated password (might need to `fg` the job to see it) in under 2 minutes (should **not** wait ~10 minutes for an overall task timeout)

## Checklist

- [x] Changes have been manually QA'd
- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-10197]: https://hpe-aiatscale.atlassian.net/browse/DET-10197?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[DET-123]: https://hpe-aiatscale.atlassian.net/browse/DET-123?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: harness/determined/deploy/local/cluster_utils.py
Comment (hamidzr) [inline harness/determined/deploy/local/cluster_utils.py:180]: what chars does this include? can we make sure it's can be picked up with a double click? eg `-` would break that flow. also maybe we printout what they'd use to create a user 

``
det -u admin u create USERNAME --password $GENRATEDPWD`

to make it easier

actually, why not prompt the user for a pwd if it's an interactive shell session? they can confirm to keep/get the autogenerated version or provide a new one

**Review by hamidzr — COMMENTED** [review]
thanks for working on this. one comment:

File: docs/release-notes/deploy-local-auto-password.rst
Comment (tara-det-ai) [inline docs/release-notes/deploy-local-auto-password.rst:8]: not sure if this is what is intended:

:orphan:

**Security Fixes**

-  CLI: If no user accounts have already been created, and neither
   ``security.initial_user_password`` in master.yaml nor the ``--initial-user-password`` is specified when
   running ``det deploy local`` with the ``master-up`` or ``cluster-up`` commands, an initial
   password will be generated and displayed to the user.

**Review by tara-det-ai — APPROVED** [review]
suggested edits

File: harness/determined/deploy/local/cluster_utils.py
Comment (jesse-amano-hpe) [reply harness/determined/deploy/local/cluster_utils.py:180]: Oh, good point -- I'll try extending this flow a little bit. I think it should still print the autogenerated password in case this is run with a closed stdin (i.e. non-interactively) or the user hits Ctrl+C or something, but prompting to change the password immediately would likely be helpful in the majority of cases. Thanks!

File: docs/release-notes/deploy-local-auto-password.rst
Comment (jesse-amano-hpe) [reply docs/release-notes/deploy-local-auto-password.rst:8]: Would you mind clarifying what the proposed change is? Just the addition of the word "the" or was there something else I missed? I would probably say "the `--initial-user-password` CLI flag" in that case, to be extra clear.

**Comment by jesse-amano-hpe** [comment]
@hamidzr I added a prompt, with the generated password still being printed as a fallback if the user can't answer the prompt correctly. Ready for re-review, I think.

File: harness/determined/deploy/local/cluster_utils.py
Comment (dannysauer) [inline harness/determined/deploy/local/cluster_utils.py:191]: This isn't guaranteed to select a password which meets the security requirements. It could generate sixteen lowercase letters, for example.

The way I've done this in the past is to select one from each set, fill the rest of the length from the combined character set, then shuffle the list by swapping each element with a random value from 0-len.

That may not matter for local deployments, but I figured it wad worth mentioning.

File: docs/release-notes/deploy-local-auto-password.rst
Comment (tara-det-ai) [inline docs/release-notes/deploy-local-auto-password.rst:8]: @jesse-amano-hpe does the system display the option to change the auto generated password? Also, we should probably clarify this is "when deploying locally" (my fault)....

e.g.,

:orphan:

**Security Fixes**

-  CLI: When deploying locally using ``det deploy local`` with ``master-up`` or ``cluster-up`` commands and no user accounts have been created yet, an initial password will be automatically generated and shown to the user (with the option to change it) if neither  ``security.initial_user_password`` in ``master.yaml`` nor the ``--initial-user-password`` CLI flag is
   present.

File: harness/determined/deploy/local/cluster_utils.py
Comment (jesse-amano-hpe) [reply harness/determined/deploy/local/cluster_utils.py:191]: That's a good point -- I'll switch it to that. Thanks.

File: e2e_tests/tests/deploy/test_local.py
Comment (carolinaecalderon) [inline e2e_tests/tests/deploy/test_local.py:259]: what does "delete-db" command have to do with this feature? is this a stray change?

File: e2e_tests/tests/deploy/test_local.py
Comment (carolinaecalderon) [inline e2e_tests/tests/deploy/test_local.py:292]: can we run this with multiple cases? of passwords that pass & those that don't & and empty case?

**Review by carolinaecalderon — COMMENTED** [review]
generally looks good, but I want to defer approval to @hamidzr since he initially reviewed the PR

File: harness/determined/deploy/local/cluster_utils.py
Comment (carolinaecalderon) [reply harness/determined/deploy/local/cluster_utils.py]: this would be great to include as additional cases to your e2e test!

File: e2e_tests/tests/deploy/test_local.py
Comment (jesse-amano-hpe) [reply e2e_tests/tests/deploy/test_local.py:259]: It's because an existing database could come with existing user/password settings (including those generated by other tests and not known to this test), which would break the tests. This gets us a clean slate.

**Comment by carolinaecalderon** [comment]
> but I want to defer approval to @hamidzr since he initially reviewed the PR

Hamid is out for the next couple weeks, so, scratch that. I'll approve once my other comments are resolved

**Pull request merged**
